### PR TITLE
Update Node.js to v16 in all RN packages

### PR DESCRIPTION
**Pull request by Pranav-yadav**
## Summary:

**NOTE**: This is a **BREAKING** change.
TLDR; Enforce minimum Node.js v16 in all RN packages.

This diff **Updates Node.js to v16** across all RN packages.

#### Context:

- For RN development and new project created; bump to node 16 was in https://github.com/facebook/react-native/pull/36217
- Recently `react-native-windows` also; updated node to v16, https://github.com/microsoft/react-native-windows/pull/11500

#### Changes:

- [BREAKING] Update Node.js to v16 across all RN packages under 'packages/' dir

## Changelog:

[GENERAL][BREAKING] - Update Node.js to v16 in all RN packages

## Test Plan:

- `yarn lint && yarn flow && yarn test-ci` --> _should be green_

### CODE REVIEW

**Comment by Pranav-yadav** [comment]
ping @cortinico, as I don't know who is taking care of RN packages ;)
PS. Sorry for the ping

**Comment by jacdebug** [comment]
Thanks for working on this, looks like this will be a breaking change, is there somewhere documented current minimum requirement for Node.js @Pranav-yadav.

**Comment by Pranav-yadav** [comment]
Hi @jacdebug,
I forgot to link to the conversation related to this diff.

Context:
https://github.com/facebook/react-native/pull/36891#issuecomment-1506869387
>Also, a general question:
what is lowest Nodejs engine compatibility provided RN packages (at least in case of codegen pkg)?
Also, for other RN packages whom should I ask this to?
>
>Edit: Actually, the main question is about ECMAScript versions compatibility :)
<br>

@cipolleschi reply to it: https://github.com/facebook/react-native/pull/36891#issuecomment-1520317237
>Hi @Pranav-yadav, you can see the engine we support from the [package.json](https://github.com/facebook/react-native/blob/main/packages/react-native/package.json#L8-L10). Apps created from the template have a similar block in their package.json

---

In-short the `package.json` of `packages/react-native` specifies `"node": ">=16";`:
https://github.com/facebook/react-native/blob/0e5d54a8ee8f43cf39c3ee9b47acdcb933a762ab/packages/react-native/package.json#L8-L10

and I think it was specified by @hoxyq during Monorepo transition, may be he can give more context.

**Comment by hoxyq** [comment]
> Hi @jacdebug, I forgot to link to the conversation related to this diff.
> 
> Context: Recently I asked following question to @cipolleschi [#36891 (comment)](https://github.com/facebook/react-native/pull/36891#issuecomment-1506869387)
> 
> > Also, a general question:
> > what is lowest Nodejs engine compatibility provided RN packages (at least in case of codegen pkg)?
> > Also, for other RN packages whom should I ask this to?
> > Edit: Actually, the main question is about ECMAScript versions compatibility :)
> 
> 
> His reply: [#36891 (comment)](https://github.com/facebook/react-native/pull/36891#issuecomment-1520317237)
> 
> > Hi @Pranav-yadav, you can see the engine we support from the [package.json](https://github.com/facebook/react-native/blob/main/packages/react-native/package.json#L8-L10). Apps created from the template have a similar block in their package.json
> 
> In-short the `package.json` of `packages/react-native` specifies `"node": ">=16";`:
> 
> https://github.com/facebook/react-native/blob/0e5d54a8ee8f43cf39c3ee9b47acdcb933a762ab/packages/react-native/package.json#L8-L10
> 
> and I think it was specified by @hoxyq during Monorepo transition, may be he can give more context.

Bump to node 16 was in https://github.com/facebook/react-native/pull/36217

**Comment by Pranav-yadav** [comment]
>Bump to node 16 was in https://github.com/facebook/react-native/pull/36217

Thanks, for pointing to the correct diff. The git blame tricked me ;)
I'll update the PR description to add that context for more accessibility.

**Comment by jacdebug** [comment]
The PR is only enforcing minimum 16 for RN development and new project created.  https://github.com/facebook/react-native/pull/36217 

When we add minimum Node.js version to all other package it will be a breaking change for all other. Given Node.js 20 is released I think this should be fine. (or is bit early to enforce it not sure? cc @robhogan)

Can you change title and desc to reflect actual change in PR, it is not just change in engine. Also update changelog to breaking. TIA

**Comment by jacdebug** [comment]
Also just checked RNW landed Update Node to v16 as well - https://github.com/microsoft/react-native-windows/pull/11500

File: package.json
Comment (jacdebug) [inline package.json:8]:
```suggestion
  "homepage": "https://reactnative.dev/",
```

File: package.json
Comment (jacdebug) [inline package.json:12]: This is not right place to add `keywords` this `package.json` is not get published to npm. It just adds noise for no value.

**Comment by Pranav-yadav** [comment]
Sure.
I'll do the required changes. That makes sense.

About, if it's too early to enforce, most of the node pkgs enforce node 16 these days. Soon, we'll start getting errors and warnings about the same from RN external deps while trying to install deps and building.

PS. Release of node 20 could be one of the reasons, but, IIRC node 20 will be LTS after Oct 2023.

File: packages/virtualized-lists/package.json
Comment (jacdebug) [inline packages/virtualized-lists/package.json:11]: I am not fan of this deep linking based on HEAD as this will break if path or title changes.

Comment (Pranav-yadav) [reply packages/virtualized-lists/package.json:11]: IIRC It is the correct way, for packages under monorepo. Though when path or title changes (rare case, as monorepo migration is completed), it'll break, may be the committers should take care of it in such scenarios.

About using HEAD based deep links, it **ensures to work** even in case we change the default branch or it's name. e.g. from `main` to `unstable`, etc.

Comment (jacdebug) [reply packages/virtualized-lists/package.json:11]: >  It is the correct way, for packages under monorepo 

Any examples? 

I see one problem documentation will diverge after published to npm by using HEAD. 
Also I noted that there is not readme for virtualized-lists, can you review all links better not add broken links.

File: package.json
Comment (Pranav-yadav) [reply package.json:8]: Sure.

File: package.json
Comment (Pranav-yadav) [reply package.json:12]: Yeah. Corrected.

File: packages/virtualized-lists/package.json
Comment (Pranav-yadav) [reply packages/virtualized-lists/package.json:11]: > > It is the correct way, for packages under monorepo
> 
> Any examples?

Currently all RN packages use HEAD based homepage links. Here:
https://github.com/search?q=repo%3Afacebook%2Freact-native+path%3Apackage.json+homepage&type=code

--
> I see one problem documentation will diverge after published to npm by using HEAD.

Afaik, something that might be not related but, since, currently packages are published manually and when it happens they are already ahead of main (in versions at least). So, using `main` based (generic) links would also cause same problems.

The optimal solution here is to change these urls during publish workflow, to include specific version based tree paths.
Which is not a difficult and shouldn't affect the current workflow.
Example: https://github.com/facebook/react-native/tree/0.72-stable/packages/virtualized-lists

--
> Also I noted that there is not readme for virtualized-lists, can you review all links better not add broken links.

Those are future proof urls, I know currently some packages which are published on npm doesn't have REDAMEs (when those are added we'll be already pointing to them), though, appending `#readme` (ie. fragment) **doesn't** break the urls.
- https://github.com/facebook/react-native/tree/HEAD/packages/virtualized-lists#readme - imho, is not a completely broken link

Comment (Pranav-yadav) [reply packages/virtualized-lists/package.json:11]: We don't need to worry about README urls.

I'll add the missing READMEs in follow up PR whenever I get some free time from my academic work and ongoing exams 😃

**Comment by Pranav-yadav** [comment]
@jacdebug Since we'll be enforcing node.js >=16, shouldn't we add CI tests to check if it builds and pass the tests on those supported versions of Node.js; for public RN packages? Or it's not necessary?

cc @cortinico / @cipolleschi

**Comment by Pranav-yadav** [comment]
>IIRC there's already a suite of tests for both the "current" nodesj version and the previous lts, test_js and test_js_prev_lts - so please make sure to upgrade the versions in the circleCI config too -> https://github.com/facebook/react-native/blob/main/.circleci/config.yml

The config uses `16.18.1` and `18.12.1` so, apart from using the LTS specified on [Node.js](https://nodejs.org/en/download/releases) ie. Gallium for `test_js_prev_lts` & Hydrogen for `test_js`, I'm not sure if there needs to any other changes?
If that's correct, let me know, so I can open a PR with bumps to respective LTS, if not; please ask someone from team to do the required changes.

PS. I'm not _too_ familiar with Circle CI configs.

File: package.json
Comment (jacdebug) [reply package.json]: This change is missing now.

File: packages/virtualized-lists/package.json
Comment (jacdebug) [reply packages/virtualized-lists/package.json:11]: Appreciated, then it is better if you can create a new PR after adding Readme with link? Updating Node.js engine is a good change after all and like to merge that.

Comment (Pranav-yadav) [reply packages/virtualized-lists/package.json:11]: Can you also comment on my latest comment below -> https://github.com/facebook/react-native/pull/37073#issuecomment-1521791948

**Comment by jacdebug** [comment]
> 

Thanks for looking into it and finding inconsistencies. Minimum requirement is correct no need to update now.

File: packages/virtualized-lists/package.json
Comment (Pranav-yadav) [reply packages/virtualized-lists/package.json]: done.

Comment (Pranav-yadav) [reply packages/virtualized-lists/package.json]: PR for adding missing READMEs and touchups to packages.json for all public RN packages: https://github.com/facebook/react-native/pull/37090

**Review by jacdebug — APPROVED** [review]
LGTM and thank you for this contribution.

**Pull request closed without merging**
